src/Datadog.Trace/HttpOverStreams/DatadogHttpClient.cs

```diff
@@ -0,0 +1,146 @@
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Datadog.Trace.HttpOverStreams.HttpContent;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.HttpOverStreams
+{
+    internal class DatadogHttpClient
+    {
+        private const int BufferSize = 10240;
+        private static readonly Vendors.Serilog.ILogger Logger = DatadogLogging.For<DatadogHttpClient>();
+
+        public async Task<HttpResponse> SendAsync(HttpRequest request, Stream requestStream, Stream responseStream)
+        {
+            await SendRequest(request, requestStream).ConfigureAwait(false);
+            return await ReadResponse(responseStream).ConfigureAwait(false);
+        }
+
+        private async Task SendRequest(HttpRequest request, Stream requestStream)
+        {
+            // TODO: Determine if it's always ASCII
+            using (var writer = new StreamWriter(requestStream, Encoding.ASCII, BufferSize, leaveOpen: true))
+            {
+                await DatadogHttpHeaderHelper.WriteLeadingHeaders(request, writer).ConfigureAwait(false);
+
+                foreach (var header in request.Headers)
+                {
+                    await DatadogHttpHeaderHelper.WriteHeader(writer, header).ConfigureAwait(false);
+                }
+
+                await DatadogHttpHeaderHelper.WriteEndOfHeaders(writer).ConfigureAwait(false);
+            }
+
+            await request.Content.CopyToAsync(requestStream).ConfigureAwait(false);
+            Logger.Debug("Datadog HTTP: Flushing stream.");
+            await requestStream.FlushAsync().ConfigureAwait(false);
+        }
+
+        private Task<HttpResponse> ReadResponse(Stream responseStream)
+        {
+            var headers = new HttpHeaders();
+            int statusCode = 0;
+            string responseMessage = null;
+
+            // hack: buffer the entire response so we can seek
+            var memoryStream = new MemoryStream();
+            responseStream.CopyTo(memoryStream);
```

Comment:
If this method is already Async (BTW the name of the method should end with the Async suffix), and we are copying all buffer to memory, any reason about this `CopyTo` being sync?